mqtt/mqtt-bridge/src/upstream/rpc/remote.rs

```diff
@@ -71,28 +112,54 @@ impl RemoteRpcHandler {
 impl EventHandler for RemoteRpcHandler {
     type Error = RpcError;
 
-    async fn handle(&mut self, event: &Event) -> Result<Handled, Self::Error> {
-        match event {
-            Event::Publication(_) => {
-                // TODO implement incoming messages translation (C2D and M2M between EdgeHubs in nested edge scenario)
-                Ok(Handled::Skipped)
+    async fn handle(&mut self, event: Event) -> Result<Handled, Self::Error> {
+        let event = match event {
+            Event::Publication(publication) if self.handle_publication(&publication).await? => {
+                return Ok(Handled::Fully);
             }
             Event::SubscriptionUpdates(subscriptions) => {
-                let mut handled = 0;
-                for subscription in subscriptions {
-                    if self.handle_subscription_update(subscription).await? {
-                        handled += 1;
+                let len = subscriptions.len();
+                match self.handle_subscriptions(subscriptions).await? {
+                    Some(skipped) if skipped.len() == len => {
+                        let event = Event::SubscriptionUpdates(skipped);
+                        return Ok(Handled::Skipped(event));
                     }
-                }
-
-                if handled == subscriptions.len() {
-                    Ok(Handled::Fully)
-                } else {
-                    Ok(Handled::Partially)
-                }
+                    Some(skipped) => {
+                        let event = Event::SubscriptionUpdates(skipped);
+                        return Ok(Handled::Partially(event));
+                    }
+                    None => return Ok(Handled::Fully),
+                };
             }
-            _ => Ok(Handled::Skipped),
-        }
+            event => event,
+        };
+
+        Ok(Handled::Skipped(event))
+    }
+}
+
+fn translate(topic_name: &str) -> Option<String> {
+    const DEVICE_OR_MODULE_ID: &str = r"(?P<device_id>[^/]+)(/(?P<module_id>[^/]+))?";
+
+    lazy_static! {
+        static ref UPSTREAM_TOPIC_PATTERNS: RegexSet = RegexSet::new(&[
+            format!("\\$iothub/{}/twin/res/(?P<params>.*)", DEVICE_OR_MODULE_ID),
+            format!(
+                "\\$iothub/{}/twin/desired/(?P<params>.*)",
+                DEVICE_OR_MODULE_ID
+            ),
+            format!(
+                "\\$iothub/{}/methods/post/(?P<params>.*)",
+                DEVICE_OR_MODULE_ID
+            )
+        ])
+        .expect("upstream topic patterns");
+    };
+
+    if UPSTREAM_TOPIC_PATTERNS.is_match(topic_name) {
+        Some(topic_name.replace("$iothub", "$downstream"))
```

Comment:
Why are we replacing with downstream?

Reply:
Per the initial design doc, events that come from upstream should have topic prefixed `$downstream`. It is done to avoid confusion when EHC doesn't know if this publication came from connected child clients or from the cloud.